modules/transactionpool/persist.go

```diff
@@ -0,0 +1,145 @@
+	// Create the persist directory if it does not yet exist.
+	err := os.MkdirAll(tp.persistDir, 0700)
+	if err != nil {
+		return err
+	}
+
+	// Open the database file.
+	tp.db, err = persist.OpenDatabase(dbMetadata, filepath.Join(tp.persistDir, dbFilename))
+	if err != nil {
+		return err
+	}
+
+	// Create the database and get the most recent consensus change.
+	cc := modules.ConsensusChangeBeginning
+	err = tp.db.Update(func(tx *bolt.Tx) error {
+		// Create the database buckets.
+		buckets := [][]byte{
+			bucketRecentConsensusChange,
+			bucketConfirmedTransactions,
+		}
+		for _, bucket := range buckets {
+			_, err := tx.CreateBucketIfNotExists(bucket)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Get the recent consensus change.
+		cc, err := tp.getRecentConsensusChange(tx)
+		if err == errNilConsensusChange {
+			return tp.putRecentConsensusChange(tx, cc)
+		}
+		return err
+	})
+	if err != nil {
+		return err
+	}
+
+	// Subscribe to the consensus set using the most recent consensus change.
+	err = tp.consensusSet.ConsensusSetSubscribe(tp, cc)
+	if err == modules.ErrInvalidConsensusChangeID {
+		// Reset and rescan because the consensus set does not recognize the
+		// provided consensus change id.
+		resetErr := tp.db.Update(func(tx *bolt.Tx) error {
+			return tp.resetDB(tx)
+		})
+		if resetErr != nil {
+			return resetErr
+		}
+		return tp.consensusSet.ConsensusSetSubscribe(tp, modules.ConsensusChangeBeginning)
+	}
+	return err
+}
+
+// getRecentConsensusChange returns the most recent consensus change from the
+// database.
+func (tp *TransactionPool) getRecentConsensusChange(tx *bolt.Tx) (cc modules.ConsensusChangeID, err error) {
+	ccBytes := tx.Bucket(bucketRecentConsensusChange).Get(fieldRecentConsensusChange)
+	if ccBytes == nil {
+		return modules.ConsensusChangeID{}, errNilConsensusChange
+	}
+	copy(cc[:], ccBytes)
+	return cc, nil
+}
+
+// putRecentConsensusChange updates the most recent consensus change seen by
+// the transaction pool.
+func (tp *TransactionPool) putRecentConsensusChange(tx *bolt.Tx, cc modules.ConsensusChangeID) error {
+	return tx.Bucket(bucketRecentConsensusChange).Put(fieldRecentConsensusChange, cc[:])
+}
+
+// transactionConfirmed returns true if the transaction has been confirmed on
+// the blockchain and false if the transaction has not been confirmed on the
+// blockchain.
+func (tp *TransactionPool) transactionConfirmed(tx *bolt.Tx, id types.TransactionID) bool {
+	confirmedBytes := tx.Bucket(bucketConfirmedTransactions).Get(id[:])
+	if confirmedBytes == nil {
+		return false
+	}
+	if confirmedBytes[0] == 1 {
+		return true
+	}
+	build.Critical("transaction database has an illegal value for a txid")
+	return false
+}
+
+// addTransaction adds a transaction to the list of confirmed transactions.
+func (tp *TransactionPool) addTransaction(tx *bolt.Tx, id types.TransactionID) error {
+	return tx.Bucket(bucketConfirmedTransactions).Put(id[:], []byte{1})
```

Comment:
I wish there was a cleaner way of creating sets (in BoltDB and Go).

Technically, you can use an empty non-nil byte slice here (`[]byte{}`). From the Bolt docs:
> It's important to note that you can have a zero-length value set to a key which is different than the key not existing.
Then you can simply return `tx.Bucket(bucketConfirmedTransactions).Get(id[:]) != nil` in `transactionConfirmed`. Checking for illegal values seems unnecessary, but that's just me.